design/design-doc-resource-claim-definition.md

```diff
@@ -0,0 +1,158 @@
+
+* https://github.com/crossplaneio/crossplane/issues/1105
+* https://github.com/crossplaneio/crossplane/issues/1106
+* https://github.com/crossplaneio/crossplane/issues/1107
+
+TODO(negz): Write this up.
+
+## Goals
+
+* Allow infrastructure operators to define claim kinds.
+* Allow a claim to be satisfied by multiple resources.
+* Surface claim to resource compatibility in Crossplane's API.
+
+TODO(negz): Write this up.
+
+## Proposal
+
+This document proposes the introduction of a new, cluster scoped, resource
+named `ResourceClaimDefinition`. A `ResourceClaimDefinition` builds on the
+existing `CustomResourceDefinition` concept, allowing an infrastructure
+operator to define arbitrary kinds of resource claim. Operators may define the
+schema for their claim, the resources it may bind to, and how it may dynamically
+provision those resources all without needing to write any code.
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceClaimDefinition
+metadata:
+  name: mysqlinstance
+spec:
+  # A controller watching for ResourceClaimDefinition uses the group, names,
+  # version, and validation fields to create a CustomResourceDefinition.
+  group: database.crossplane.io
+  names:
+    kind: MySQLInstance
+    listKind: MySQLInstanceList
+    plural: mysqlinstances
+    singular: mysqlinstance
+  # We do not support multiple versions of the same ResourceClaimDefinition.
+  version: v1alpha1
+  # The validation field is used as an overlay when creating the CRD associated
+  # with this RCD. The base layer would contain all standard claim fields such
+  # that the operator did not need to respecify them all.
+  validation:
+    openAPIV3Schema:
+      description: A MySQLInstance is a portable resource claim that may be satisfied
+        by binding to a MySQL managed resource such as an AWS RDS instance or a GCP
+        CloudSQL instance.
+      properties:
+        spec:
+          properties:
+            engineVersion:
+              description: EngineVersion specifies the desired MySQL engine version,
+                e.g. 5.7.
+              enum:
+              - "5.6"
+              - "5.7"
+              type: string
+  # Satisfaction specifies one or more "resource groups" that can satisfy this
+  # claim kind. Typically each group would correspond to one or more resources
+  # from a single cloud provider, but you could also mix and match resources
+  # from different providers (e.g. a GKE cluster and an S3 bucket) if that was
+  # somehow useful to you.
+  satisfaction:
+  # GCP is a resource group that can satisfy this claim with a single resource.
+  - name: GCP
+    description: A MySQLInstance can be satisfied by a CloudSQLInstance.
+    resources:
+    - resource:
+        apiVersion: database.gcp.crossplane.io/v1beta1
+        kind: CloudSQLInstance
+      class:
+        apiVersion: database.gcp.crossplane.io/v1beta1
+        kind: CloudSQLInstanceClass
+        strategies:
```

Comment:
> Have you thought at all yet about how to accommodate more strategies? 

I imagine a strategy to be two things:

1. Fields on a `ResourceClaimSatisfaction` CR that configure a particular strategy.
1. Code in the `ResourceClaimSatisfaction` controller(s) that enact the strategy.

So therefore accommodating new strategies would involve adding fields, and releasing new controller code.

Reply:
> My guess would be that someone at some point will probably want a complex strategy, and may want to use a familiar resource engine (e.g. helm, kustomize, kapitan).

I'm not convinced that such tools are a good fit (or bring much value) for the dynamic provisioning use case, so my thinking is we should focus on supporting a set of targeted, useful strategies rather than attempting to deliver an open ended template-stacks-like experience in which you can bring whatever templating engine you're familiar with.

While dynamic provisioning is template-ish (we take a claims and classes as inputs, and produce managed resources as outputs) there are also a lot of constraints on the process. We require claims, classes, and managed resources to have certain shapes, act certain ways, and support certain patterns. We probably only want to use the class `.specTemplate` and a subset of the claim `.spec.` as inputs, and only produce a subset of the managed resource's `.spec` as an output, leaving the rest of the object's values (name, external name annotation, binding phase, references, etc) up to the opinions of the dynamic provisioning controller.